odoo/addons/base/models/ir_ui_menu.py

```diff
@@ -74,62 +74,69 @@ def _check_parent_id(self):
-
-        # first discard all menus with groups the user does not have
         group_ids = set(self.env.user._get_group_ids())
         if not debug:
-            group_ids = group_ids - {self.env['ir.model.data']._xmlid_to_res_id('base.group_no_one', raise_if_not_found=False)}
-        menus = menus.filtered(
-            lambda menu: not (menu.groups_id and group_ids.isdisjoint(menu.groups_id._ids)))
+            group_ids.discard(self.env['ir.model.data']._xmlid_to_res_id('base.group_no_one', raise_if_not_found=False))
+
+        # retrieve menus with a domain to filter out menus with groups the user does not have.
+        # It will be used to determine which ones are visible
+        menus = self.with_context({}).search_fetch(
+            # Don't use 'any' operator in the domain to avoid ir.rule
+            ['|', ('groups_id', '=', False), ('groups_id', 'in', tuple(group_ids))],
+            ['parent_id', 'action'], order='id',
+        ).sudo()
 
         # take apart menus that have an action
-        actions_by_model = defaultdict(set)
+        action_ids_by_model = defaultdict(list)
         for action in menus.mapped('action'):
             if action:
-                actions_by_model[action._name].add(action.id)
-        existing_actions = {
-            action
-            for model_name, action_ids in actions_by_model.items()
-            for action in self.env[model_name].browse(action_ids).exists()
-        }
-        action_menus = menus.filtered(lambda m: m.action and m.action in existing_actions)
-        folder_menus = menus - action_menus
-        visible = self.browse()
+                action_ids_by_model[action._name].append(action.id)
 
-        # process action menus, check whether their action is allowed
-        access = self.env['ir.model.access']
         MODEL_BY_TYPE = {
             'ir.actions.act_window': 'res_model',
             'ir.actions.report': 'model',
             'ir.actions.server': 'model_name',
         }
+        def exists_actions(model_name, action_ids):
+            """ Return existing actions and fetch model name field if exists"""
+            if model_name not in MODEL_BY_TYPE:
+                return self.env[model_name].browse(action_ids).exists()
+            records = self.env[model_name].sudo().with_context(active_test=False).search_fetch(
+                [('id', 'in', action_ids)], [MODEL_BY_TYPE[model_name]], order='id',
+            )
+            if model_name == 'ir.actions.server':
+                # Because it is computed, `search_fetch` doesn't fill the cache for it
+                records.mapped('model_name')
+            return records
 
-        # performance trick: determine the ids to prefetch by type
-        prefetch_ids = defaultdict(list)
-        for action in action_menus.mapped('action'):
-            prefetch_ids[action._name].append(action.id)
+        existing_actions = {
+            action
```

Comment:
Maybe make a set of ids instead of recordsets.
`for action_id in exists_actions(model_name, action_ids).ids`

Reply:
I thought about it, but I prefer to keep a record here because `action` have different type of model and I don't want only compare `id` (`action in existing_actions`). Your suggestion is normally correct since all type of action share the same `id` but I don't like to rely on that, I found it more brittle (also the performance gain was minimal). What do you think ?

Reply:
As you wish.
I suggested this because you only check the existence.